### ISSUE TITLE
Bugs in the UI clicking interface

### ISSUE THREAD

**Issue by cseager** [jmoenig/Snap--Build-Your-Own-Blocks#1067]
Today a bunch of new bugs appeared in the interface. When creating a custom block, I can't click on dropdown lists such as "set variable to...". When I try to click on it to select a variable, the mouse just picks up the block instead... I'm being forced to use all global variables because I can't select script variables due to this bug. I also can't seem to pick up some of my custom blocks out of the palette area and into the scripting area... no matter where I click, the block just executes and causes a type error. Also I can't save, because of a type error somewhere in the project. Clicking on the stop button in the tool bar to stop all scripts from executing does not solve the problem. Here's a screenshot of the error I get when trying to save. Deleting the custom block allows me to save. 

<img width="446" alt="screen shot 2015-12-15 at 5 24 23 pm" src="https://cloud.githubusercontent.com/assets/2761922/11826109/3f09d4fc-a351-11e5-85d1-b9c008fcd7e6.png">

**Comment by cseager** [comment]
update: I downloaded the repo and switched the HEAD to 42fd60427f8224311a613426b9589c5d7791c940. That version seems to be working as expected without problems.

**Comment by jmoenig** [comment]
Please make sure to flush your cache until the module versions show the following:

![404151215](https://cloud.githubusercontent.com/assets/2596285/11826663/1cfdb9b0-a387-11e5-8887-bc9931e8a59c.png)

Let me know if the problems still persist after that.
Thanks!

**Comment by cseager** [comment]
@jmoenig ever since I loaded SNAP from my local rolled-back repo yesterday, I have not been able to reproduce some of the original bugs. The module versions are as you show when I load it from the web now. I'm pretty confused about what's going on here... I had no problems working over the weekend, and had cleared my cache out a couple of weeks ago when I got a login error. So I think it's pretty likely that I was working with up-to-date modules until yesterday, and I'm not sure in what way they could go wrong to cause this. 

I was able to reproduce some other crashing behavior I got yesterday. When you click on the following test block without inputting a list, it basically crashes the browser tab to the point where the only way to stop is to close the tab. I can't click on the block, I can't click the stop button, I can't even reload the page at that point. Raising a TypeError was the previous behavior. 
![image](https://cloud.githubusercontent.com/assets/2761922/11848109/bd553d22-a3ee-11e5-8e13-aa252431bcdf.png)

Here's a link to a test program. Just click the test block to execute it to see what I mean. 
http://snap.berkeley.edu/snapsource/snap.html#present:Username=cseager&ProjectName=test

**Comment by brianharvey** [comment]
You have to clear your cache every time there's a new Snap<i>!</i> release.  Having done it weeks ago doesn't help!  ☺

**Comment by cseager** [comment]
@brianharvey the reason this confuses me is that I experimented with that exact thing locally. I cleared my cache, set my HEAD to the Nov 19th commit, loaded the page, did some stuff, then set my HEAD to the latest commit. It did not break and the modules loaded correctly without clearing my cache. Wouldn't I be able to break it by doing that locally, or is there something that causes chrome to not cache it when I'm just hosting it on my computer? 

Also, if this is expected behavior, then it would be really nice if the program would TELL students and teachers when they need to clear their cache, or if that isn't possible at least announce a new release on snap.berkeley.edu so that we have some way of knowing what might be happening. 

The crashing bug in my previous comment still stands as a problem though, and a pretty bad one since it would cause a student to lose work if they accidentally clicked on a block that should raise a TypeError.

**Comment by cycomachead** [comment]
I can confrim the infinite loop is a problem!

Chrome still caches local files, though it sounds like you've cleared the cache there too...

Do you still notice the UI bugs? What about the save issues? If you have a project for that, it'd be helpful to test.

**Comment by jmoenig** [comment]
Thanks for the bug report about the LENGTH reporter. I've already fixed this in my dev version, and it will be in the next release.

**Comment by jmoenig** [comment]
... but this isn't a new problem, it has been in all prior version of Snap. We'll fix it in the next release.

**Comment by cycomachead** [comment]
> ... but this isn't a new problem, it has been in all prior version of Snap. We'll fix it in the next release.

Before `length of` would throw an error, right?

**Comment by cseager** [comment]
Yes, the length reporter bug is new. I double checked the older version and it correctly throws an error when not given a list, which is not a fatal killer for the whole tab. Now, it does NOT throw an error, and just gets caught in an infinite loop that I can't find any way to stop, which is worse because that can mean any unsaved work is not salvageable.

**Comment by cseager** [comment]
@cycomachead correct, I cleared my cache multiple times. Many of the UI bugs disappeared after that. The save issue is unfortunately not saved... because I couldn't find a way to save it, and unfortunately at that moment I was more focused on getting materials ready for my students than on preserving details for a bug report :(

**Comment by jmoenig** [comment]
Can you tell us more about the save issue? Were you able to save this project in the old version? Can you share the version that's currently saved in the cloud? Tnx!

**Comment by brianharvey** [comment]
I am a git novice, but I think the whole point of git is to make sure you have a consistent version on your computer, right?  The problem about stuff in the cache is ending up with some files from 4.0.4 and some from 4.0.3.  So it's different from backing up in git.  Or maybe I'm not understanding what you did.  :-/

**Comment by cycomachead** [comment]
> I am a git novice, but I think the whole point of git is to make sure you have a consistent version on your computer, right?

Well, yes, so it mirrors whatever he checked out, but the browser can still cache a different version of a file, regardless of what's in the file system.

**Comment by cseager** [comment]
@jmoenig sorry, I actually abandoned my project and started over from scratch, cloning the git repo and reverting it to an earlier version, when I ran into that issue. I've been trying to reproduce it but I'm sure it will be impossible to figure out exactly what I had cached in my browser at that point... so next time I run into an issue I will know about gathering module version information. I will go ahead and close this issue since @cycomachead made a separate issue about the infinite looping problem. 

@brianharvey what I did was 1) Clear my cache completely. 2) Revert my local files to an earlier version of the program. This is done in git by moving the `HEAD` tag to a commit of your choice... the files on the system will reflect whatever that version is. Returning to the current version is as simple as changing the `HEAD` tag again. Image below is a picture of what I'm doing. 3) Test SNAP by opening the local index.html file from my computer. This causes all of the modules to load as they were 3 days ago, and this way I can confirm that the infinite loop bug wasn't an issue back then. 4) Return the `HEAD` tag to the most recent commit. This causes all of the modules to load to their most recent versions, so maybe chrome doesn't cache for files on my computer? I'm not super familiar with how caching works on browsers. 
<img width="1142" alt="screen shot 2015-12-16 at 6 48 37 pm" src="https://cloud.githubusercontent.com/assets/2761922/11857683/705e7c56-a428-11e5-8e75-69985d84c1af.png">
